surfactant/filetypeid/id_extension.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2023 Lawrence Livermore National Security, LLC
+# See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: MIT
+import pathlib
+import re
+from typing import Optional
+
+from loguru import logger
+
+import surfactant.plugin
+
+
+@surfactant.plugin.hookimpl
+def identify_file_type(filepath: str) -> Optional[str]:
+    # pylint: disable=too-many-return-statements
+    _filetype_extensions = {
+        ".sh": "SHELL",
+        ".bash": "BASH",
+        ".zsh": "ZSH",
+        ".py": "PYTHON",
+        ".pyc": "PYTHON_COMPILED",
+        ".js": "JAVASCRIPT",
+        ".css": "CSS",
+        ".html": "HTML",
+        ".htm": "HTML",
+        ".php": "PHP",
+    }
+    _interpreters = {
+        b"sh": "SHELL",
+        b"bash": "BASH",
+        b"zsh": "ZSH",
+        b"php": "PHP",
+        b"python": "PYTHON",
+        b"python3": "PYTHON",
+    }
+    try:
+        with open(filepath, "rb") as f:
+            head = f.read(256)
+            if head.startswith(b"<!DOCTYPE html>"):
+                return "HTML"
+            if head.startswith(b"#!") and b"\n" in head:
+                end_line = head.index(b"\n")
+                head = head[:end_line]
+                for interpreter, filetype in _interpreters.items():
+                    if re.search(interpreter, head):
```

Comment:
Any idea how the performance is for `re.search` vs just using the `in` keyword?

Reply:
Okay, so I put together some quick scripts to test with `time python3 <script>`. I got around 2.3s using `re.search`, and 1.2s using `in`.

```python
import re

interpreters = {
    b"sh": "SHELL",
    b"bash": "BASH",
    b"zsh": "ZSH",
    b"php": "PHP",
    b"python": "PYTHON",
    b"python3": "PYTHON",
}

candidate = b"#!/bin/fsdafsdf/local/bas/bas/bas/bas/pytho/python3"

for i in range(0, 1000000):
    for interpreter, filetype in interpreters.items():
        if re.search(interpreter, candidate):
            pass

```

```python
interpreters = {
    b"sh": "SHELL",
    b"bash": "BASH",
    b"zsh": "ZSH",
    b"php": "PHP",
    b"python": "PYTHON",
    b"python3": "PYTHON",
}

candidate = b"#!/bin/fsdafsdf/local/bas/bas/bas/bas/pytho/python3"

for i in range(0, 1000000):
    for interpreter, filetype in interpreters.items():
        if interpreter in candidate:
            pass
```

Out of curiosity, I tried out using Aho-Corasick (on strings though... using the pyahocorasick library) to see how fast it could be if we really wanted to get a faster search; with this, the test took 0.35s to run:

```python
import ahocorasick

automaton = ahocorasick.Automaton()

interpreters = {
        "sh": "SHELL",
        "bash": "BASH",
        "zsh": "ZSH",
        "php": "PHP",
        "python": "PYTHON",
        "python3": "PYTHON3",
}

for interpreter, filetype in interpreters.items():
    automaton.add_word(interpreter, filetype)

automaton.make_automaton()

candidate = "#!/bin/fsdafsdf/local/bas/bas/bas/bas/pytho/python3"

for i in range(0, 1000000):
    for item in automaton.iter_long(candidate):
        pass
```

Reply:
At some point, it might be nice to switch some of the different file type ID checks over to a faster check -- but saving a second or two isn't that much compared to the time it takes to generate hashes and analyze large binaries.

Reply:
Okay, so I put together some quick scripts to test with `time python3 <script>`. I got around 2.3s using `re.search`, and 1.2s using `in`.

```python
import re

interpreters = {
    b"sh": "SHELL",
    b"bash": "BASH",
    b"zsh": "ZSH",
    b"php": "PHP",
    b"python": "PYTHON",
    b"python3": "PYTHON",
}

candidate = b"#!/bin/fsdafsdf/local/bas/bas/bas/bas/pytho/python3"

for i in range(0, 1000000):
    for interpreter, filetype in interpreters.items():
        if re.search(interpreter, candidate):
            pass

```

```python
interpreters = {
    b"sh": "SHELL",
    b"bash": "BASH",
    b"zsh": "ZSH",
    b"php": "PHP",
    b"python": "PYTHON",
    b"python3": "PYTHON",
}

candidate = b"#!/bin/fsdafsdf/local/bas/bas/bas/bas/pytho/python3"

for i in range(0, 1000000):
    for interpreter, filetype in interpreters.items():
        if interpreter in candidate:
            pass
```

Out of curiosity, I tried out using Aho-Corasick (on strings though... using the pyahocorasick library) to see how fast it could be if we really wanted to get a faster search; with this, the test took 0.35s to run:

```python
import ahocorasick

automaton = ahocorasick.Automaton()

interpreters = {
        "sh": "SHELL",
        "bash": "BASH",
        "zsh": "ZSH",
        "php": "PHP",
        "python": "PYTHON",
        "python3": "PYTHON3",
}

for interpreter, filetype in interpreters.items():
    automaton.add_word(interpreter, filetype)

automaton.make_automaton()

candidate = "#!/bin/fsdafsdf/local/bas/bas/bas/bas/pytho/python3"

for i in range(0, 1000000):
    for item in automaton.iter_long(candidate):
        pass
```

At some point, it might be nice to switch some of the different file type ID checks over to a faster method (magic bytes, srec/hex, etc) -- but saving a second or two isn't that much compared to the time it takes to generate hashes and analyze large binaries.